include/aws/nitro_enclaves/kms.h

```diff
@@ -172,6 +172,120 @@ struct aws_kms_decrypt_response {
     struct aws_allocator *const allocator;
 };
 
+struct aws_kms_encrypt_request {
+    /**
+     * Plaintext to be encrypted.
+     *
+     * Required: Yes.
+     */
+    struct aws_byte_buf plaintext;
+
+    /**
+     * Specifies the encryption algorithm that will be used to decrypt the ciphertext.
+     * Specify the same algorithm that was used to encrypt the data.
+     * If you specify a different algorithm, the Decrypt operation fails.
+     *
+     * Required: No.
+     */
+    enum aws_encryption_algorithm encryption_algorithm;
+
+    /**
+     * Specifies the encryption context to use when decrypting the data. An encryption
+     * context is valid only for cryptographic operations with a symmetric CMK.
+     * The standard asymmetric encryption algorithms that AWS KMS uses do not
+     * support an encryption context.
+     *
+     * An encryption context is a collection of non-secret key-value pairs that
+     * represents additional authenticated data. When you use an encryption context
+     * to encrypt data, you must specify the same (an exact case-sensitive match)
+     * encryption context to decrypt the data. An encryption context is optional
+     * when encrypting with a symmetric CMK, but it is highly recommended.
+     *
+     * Required: No.
+     */
+    struct aws_hash_table encryption_context;
+
+    /**
+     * A list of grant tokens.
+     *
+     * For more information, see
+     * <a href="https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#grant_token">Grant Tokens</a>
+     * in the AWS Key Management Service Developer Guide.
+     *
+     * Required: No.
+     */
+    struct aws_array_list grant_tokens;
+
+    /**
+     * Specifies the customer master key (CMK) that AWS KMS will use to decrypt
+     * the ciphertext. Enter a key ID of the CMK that was used to encrypt the ciphertext.
+     *
+     * If you specify a KeyId value, the Decrypt operation succeeds only if the specified
+     * CMK was used to encrypt the ciphertext.
+     *
+     * This parameter is required only when the ciphertext was encrypted under an asymmetric CMK.
+     * Otherwise, AWS KMS uses the metadata that it adds to the ciphertext blob to determine
+     * which CMK was used to encrypt the ciphertext. However, you can use this parameter to
+     * ensure that a particular CMK (of any kind) is used to decrypt the ciphertext.
+     *
+     * Required: No.
+     */
+    struct aws_string *key_id;
+
+    /**
+     * Recipient field.
+     *
+     * Required: No.
+     */
+    struct aws_recipient *recipient;
```

Comment:
The Recipient field is only supported on the 3 APIs available in this repo: Decrypt, GenerateRandom and GenerateDataKey, since they are the ones which might use KMS to gain privileged access. On Encrypt, this is not required, as the enclave already has the confidential data and ensures that the connection to KMS is trusted via TLS. I believe AWS KMS will just ignore the Recipient field set here, as it's not part of the REST API.

Reply:
I see, thank you for clarification!
Is it only these 3 (forever or for now?) and if not, is there a document that lists APIs which support recipient attestations (currently, this official document: https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html doesn't show it)?

My use case needs are
1. ensuring encrypted data origin: if Encrypt was supported, my assumption is that one could rely on AWS KMS key policy (~> if only a particular enclave can call Encrypt with a particular key, then data encrypted with that key originated from that enclave). If that's not possible, I guess the alternative would be to attest-sign on the plaintext data, bundle the attestation payload with the plaintext data for encryption (either generating a dedicated keypair and putting the pubkey in userdata for NSM attestation request; or putting the plaintext in userdata), and verify it after decryption. Is that correct?

2. ensuring encrypted data "freshness": I guess one way would be to get a timestamp or a counter via TLS from some trusted service, and bundling that in plaintext and checking that after decryption. Or is there a way to retrieve something like this from NSM (judging from https://github.com/aws/aws-nitro-enclaves-nsm-api/blob/main/nsm-io/src/lib.rs#L78 my guess is no?)? One alternative here would be if one can restrict REST API "write" actions on external persistent storage services (S3, DynamoDB, ...) to require certain recipient attestation values, similarly to currently AWS KMS key policies -- is that supported?

3. hassle with vsock: many AWS API libraries hardcode TCP at this moment, so it's a bit of dependency patching rabbit hole

Given the last point and potential plans for support in other APIs, would you suggest to add it here (and perhaps remove the recipient for now) or should I maintain a separate fork?

Reply:
Yes, there is some holdup with updating the API documentation. However, I don't see why this patch shouldn't be included, as long as it doesn't contain the Recipient field.

1. Putting the Plaintext data in the Attestation Document userdata and encrypting that should be ok, but there is a limit on how large the userdata can be: the whole request structure cannot exceed 4kiB, so the user provided fields are limited to around 3kiB total. Another option would be to use GenerateDataKey to get an encryption key for the data and place the ciphertext of the key in userdata instead, and bundle the attestation document generated this way with the encrypted data. It's a good way to prove where the data originated, even stronger than that based on the Encrypt API, since KMS policies can be changed after the fact. There's a bit of work needed to interact with the Attestation document more directly after it was generated (https://github.com/aws/aws-nitro-enclaves-nsm-api/issues/4) that should help with this.

2. The Attestation Document contains the UTC timestamp at the moment of generation, but the enclave also has access to the same trusted clock source. Only KMS supports verifying the Attestation Document at the moment.

3. I agree that this can be a sore point at the moment and there's definitely a lot of room for improvement; the vision was that the enclave itself would only use a small series of trusted services that implement Attestation (like AWS KMS) and have the parent do the rest, such as retrieving or uploading data from S3/Dynamo. This way the enclave contains a smaller Trusted Computing Base and there are fewer things that might go wrong, which limits data exfiltration potential and attack surface; keeping complexity down also helps with reliability and debugging, given that visibility inside the enclave is not great in production. We don't know yet if this is realistic yet. We are planning on adding similar SDKs for at least a select number of languages.

Reply:
ok, thank you for your reply! I've removed the recipient field